### PR TITLE
feat(pnp): no longer virtualize single instance packages

### PR DESCRIPTION
**Pull request by andreialecu**
Keeping as draft for now.

**What's the problem this PR addresses?**

All packages with peer dependencies were being virtualized, adding `$$virtual` to their paths. This is unnecessary when only one instance of a package is needed.

This creates developer experience issues:
1) Stack traces become unnecessarily confusing, showing lots of `$$virtual` paths.
2) VSCode `Go to Definition` functionality across workspaces may go to a path under the `$$virtual` file system instead of the real file. This can cause issues with the TypeScript Server, and general developer confusion.

**How did you fix it?**

Added another pass after peer dependency resolution which turns virtual packages that have a single instance back into normal packages.

Removed dead/obsolete code which when fixed resulted in bugs and will soon be removed: https://github.com/yarnpkg/berry/pull/2066#issuecomment-719764355

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-pnp/sources/AbstractPnpInstaller.ts [hunk 1]
```diff
@@ -86,7 +78,7 @@ export abstract class AbstractPnpInstaller implements Installer {
       buildScripts.length = 0;
     }
 
-    const packageFs = !hasVirtualInstances && pkg.linkType !== LinkType.SOFT
+    const packageFs = pkg.linkType !== LinkType.SOFT
```
Comment (arcanis) [inline packages/plugin-pnp/sources/AbstractPnpInstaller.ts:81]: This change means that all packages with peer dependencies will be unplugged once more (so even if there is a single instance, there will be two unplugs).

File: packages/yarnpkg-core/sources/Project.ts [hunk 2]
```diff
@@ -2122,8 +2122,64 @@ function applyVirtualResolutionMutations({
...
+    // find any workspaces that only have one instance
+    for (const descriptor of allDescriptors.values()) {
+      const resolution = allResolutions.get(descriptor.descriptorHash);
+      if (!resolution) continue;
+
+      const virtualPackage = allPackages.get(resolution);
+      if (!virtualPackage) continue;
+
+      if (structUtils.isVirtualLocator(virtualPackage)) {
+        packagesWithInstanceCount.set(virtualPackage.identHash, {
```
Comment (arcanis) [inline packages/yarnpkg-core/sources/Project.ts:2137]: This doesn't work, you can have multiple versions of a single package that lists peer dependencies; since you're only keying on `identHash` (rather than `locatorHash` from the devirtualized locator), you will collapse them all into a single dict entry.

**Review by arcanis — COMMENTED** [review]
As mentioned on Discord, I'm not in favour of this change at the moment because:

- it would have the negative impact that it would make it harder for users to understand under which conditions would a package be virtualized.

- it would make it *easier*, but wouldn't address the underlying problem. As a result, a project would work just fine until you add an unrelated dependency and suddenly things break in another package. The right fix is to behave properly with virtual packages, not to avoid them.

- it makes the code ever so slightly more complicated, and as you saw the virtual code is already plenty complex already! additional logic needs to have a high value.

- I have the feeling that the fix for your use case may just be to devirtualize the paths before sending them to VSCode, which would help with editing files at the cost of a slight correctness in the import resolution, most likely harmless.

File: packages/plugin-pnp/sources/AbstractPnpInstaller.ts (hunk 1 above)
Comment (andreialecu) [reply packages/plugin-pnp/sources/AbstractPnpInstaller.ts:81]: If this is true, then this is a bug that has existed for a long time. 

`hasVirtualInstances` is always `false` so it was dead code.

File: packages/yarnpkg-core/sources/Project.ts (hunk 2 above)
Comment (andreialecu) [reply packages/yarnpkg-core/sources/Project.ts:2137]: Yes, but that is correct. I want to key on `identHash` to count the number of instances an `identHash` has. If there is more than one, it isn't affected.

**Comment by andreialecu** [comment]
Hey @arcanis 👋

I only opened this and kept it as draft to run the test suite. I realize it has little chance of being merged.

As a counter-argument to the points you mentioned:

* I have noticed things breaking simply from adding a peer dependency: #1932. The negative impact you mentioned already happened when I hit that. This simply moves it slightly "later", instead of hitting it "earlier".
* If more consistency is the goal, perhaps everything should use virtual paths regardless of whether they have peers or not. Then there would be no surprises at all. 🙂 
* Devirtualizing single instance packages simplifies the resulting instalation quite a bit, as can be seen from the `.pnp.js` file. The code this PR adds is pretty straight forward and can be made even more readable if needed.

In practice, getting virtual paths to work right in all cases in VSCode might not be possible without full official integration with TypeScript, so this is an interim solution.

Note that it can be limited to just workspaces. I have just enabled it for all packages as an experiment: [`c1bd9af` (#2070)](https://github.com/yarnpkg/berry/pull/2070/commits/c1bd9af8dd1593c7aa9f68f51237280a93cc26a6)

File: packages/yarnpkg-core/sources/Project.ts (hunk 2 above)
Comment (andreialecu) [reply packages/yarnpkg-core/sources/Project.ts:2137]: Unless you mean keying by `locatorHash` would result in additional packages being correctly devirtualized and affected by this change? 

Things that with `identHash` it currently misses?

Comment (andreialecu) [reply packages/yarnpkg-core/sources/Project.ts]: Yup, seems that was the case. With `identHash` it was more conservative (but still safe). `locatorHash` catches way more packages to be devirtualized.

**Comment by larixer** [comment]
For the record, I think reducing the number of virtual paths used in a dependency graph representation seems like a good direction for me. The rationale is simple - the virtual paths is a kind of magic and we want to keep the amount of magic to an absolute minimum possible.

**Comment by andreialecu** [comment]
Just a note on developer experience:

Here's a random error using current behavior:

image

Here's the same error using the proposed behavior:

image

**Comment by andreialecu** [comment]
A bunch of edge-cases are broken when using the changes in this PR, so I don't recommend using it for now.

Fixing the tests would be the first step, but I don't have much time to dig into it right now.

**Comment by arcanis** [comment]
I'm going to close this PR since we don't intend to merge it, due to having some heavy side effects in how our core is made (and I'm trying to clean a bit the PR dashboard).

If that can help, after thinking about this a bit more, I think a simpler approach would be to *not* prevent virtualization, but instead to devirtualize them at the linker stage. This way you'd scope your changes to a linker, without having to change the rest of the software's assumptions.

Still, this should live in a plugin, I don't think it would be a good idea to do that in the core right now (especially now that we have implemented #2136).
